### ISSUE TITLE
Hyperband bracket assignment using hash in parallelization

### ISSUE THREAD

**Issue by donny8** [optuna/optuna#3083]
I think there could be a problem when using Hyperband pruner in parallelization.
In Hyperband, each trial is assigned to the bracket by computing hash("{}{}".format(study.studyname, trial.number).
However, the hash gives different value for the same input when executed on different terminal.
A trial n, for example, could be assigned to the bracket 1 at process 1 and the bracket 2 at process 2 , which ruins the benefits of using Hyperband pruner instead of the SuccesiveHalving pruner.

## Expected behavior

A trial needs to be assigned to the identical bracket among each process.
 

## Environment

- Optuna version: 2.10.0
- Python version: 3.6.9
- OS: Linux

## Error messages, stack traces, or logs

I attached an custom-made log that I'd added in the SuccessiveHalving code, which is a part of the Hyperband.
The trials 1, 2, 3, 4 were included in bracket 1 at both process 1 and 2
However, the trials 5, 7, 9 were included in the bracket only at process 2. 

**Trials for bracket 1 at process 1**
![image](https://user-images.githubusercontent.com/55653250/140302703-a7ea8e8a-0f9d-4be1-ae1c-246686535c17.png)

**Trials for the same bracket 1 at process 2**
![image](https://user-images.githubusercontent.com/55653250/140302720-94e346e1-0984-4278-a278-005534efe9cc.png)

## Steps to reproduce

1. Add the following codes at the beginning of the **def _get_competing_values( )** in the **_successive_halving.py**
```
for t in trials :
    if rung_key in t.system_attrs:
        compet = [t.system_attrs[rung_key]]
        print('   Prune SH Get_competing_values: '+ rung_key, ' trial ',t.number,'competing ',compet)
```
2. Check if the trials consisting of a bracket differs among each process

## Additional context (optional)

As you can see [here](https://docs.python.org/3.3/using/cmdline.html), it was due to the update that hash randomization is turned on after python version 3.3.
Of course, the randomization could be turned off by setting the PYTHONHASHSEED=0 when executing the python script.
Since I did not see any warning or suggestion about this problem, I would like to ask if my assertion was legit.

**Comment by hvy** [comment]
I think you're right. It's clearly manifested when creating the bracket studies (a study which contains a subset of the trials) in https://github.com/optuna/optuna/blob/v2.10.0/optuna/pruners/_hyperband.py.

Let me share a somewhat more limited but self contained example that demonstrates the issue that you're describing.

Run the following script from more than one process and notice that the printed brackets are not the same. SQLite is required.

```python
from collections import defaultdict

import optuna as ot


def objective(t):
    x = t.suggest_float("x", 0, 1)

    for step in range(10):
        t.report(step=step, value=x * step)
        if t.should_prune():
            raise ot.TrialPruned

    return x


if __name__ == "__main__":
    study = ot.create_study(
        study_name="test_hyperband_3083",
        storage="sqlite:///test_hyperband_3083.db",
        pruner=ot.pruners.HyperbandPruner(),
        load_if_exists=True,
    )
    study.optimize(objective, n_trials=10)

    # Check bracket splits. Using private APIs, only use for testing.
    brackets_to_trial_numbers = defaultdict(list)
    for t in study.trials:
        brackets_to_trial_numbers[study.pruner._get_bracket_id(study, t)].append(t.number)
    print(dict(brackets_to_trial_numbers))
```

One approach is to simply replace `hash` with a deterministic algorithm from `hashlib`.

**Comment by donny8** [comment]
I'd executed the script above and assured that trials in the brackets keep changing in parallelization.
I would also consider your suggestion to use `hashlib` instead of `hash`
Thank you very much, I really appreciate it.

**Comment by hvy** [comment]
Thanks for the prompt response.
How about keeping the issue open until it's been fixed?

**Comment by donny8** [comment]
Oh I thought either using `PYTHONHASHSEED=0` or `hashlib` would be an adequate solution for the issue.
I would keep opening this issue until it's been fixed.

**Comment by avats-dev** [comment]
From what I understand, https://github.com/optuna/optuna/blob/40b097650d312913989b5825aa9c9adf4b15c396/optuna/pruners/_hyperband.py#L238-L260 is the relevant code and we need to change `hash` function here as it is randomizing based on what @donny8 mentioned above, right?

**Comment by donny8** [comment]
What you had understood is absolutely correct!

**Comment by avats-dev** [comment]
Ok, so I am going to try to resolve this one and raise a PR. I'll reach out if I got stuck somewhere. Thanks 🎉

**Comment by avats-dev** [comment]
Ok, this could be resolved by using any particular algorithm from `hashlib` but which one should I select and also, as there are no major concerns of collisions and instead we need consistency, wouldn't it be better to just set `PYTHONHASHSEED=0` ? But, I am not sure about where should this be set in the code, as if in class definition or this particular method or globally.

**Comment by donny8** [comment]
In my case, I am using `PYTHONHASHSEED=0`.
You can either put it in the code or outside the code

PYTHONHASHSEED inside the code : 

- refer this [link](https://gist.github.com/mkolod/853cda9950b898d056ac149abc45417a)

PYTHONHASHSEED outside the code :  

- Add the phrase before the "python main.py". Ex) "PYTHONHASHSEED=0 python main.py"

I did not the algorithms in `hashlib` since I did not want to change the base codes.

**Comment by hvy** [comment]
Using `PYTHONHASHSEED` is maybe a reasonable workaround, in user-land. As for how to address that in Optuna, I suggest we adopt some hashing algorithm in `hashlib` (we don't want to cause environment variable conflicts with the user and as stated above, this variable must be set prior to the Python process execution). I'm not sure which hashing algorithm but I think any naive one that's not too slow will do. Perhaps start with sha256, or similar. Open for comments.

**Comment by toshihikoyanase** [comment]
@hvy I took a brief benchmark using `timeit`.

The result shows that `hashlib.sha*` functions were 7 to 8 times slower than `hash()`.  This is mainly because `hashlib.sha*` returns a byte array and we need to convert it to an integer to calculate the remainder divided by `self._total_trial_allocation_budget`. @c-bata suggested an alternative function `binascii.crc32` , and it seems a bit slower than `hash`, but it is much faster than `hashlib.sha*`. I think we don't need cryptographic hash functions for this purpose, and `binascii.crc32` may be a reasonable choice among them.

Alternatively, we may just take `trial.number % self._total_trial_allocation_budget` as suggested in https://github.com/optuna/optuna/pull/809#discussion_r361076742. I'm not sure the randomness of `hash` functions affects the search performance.

| name | time (sec.) |
|  ---     | ---             |
| `hash` |  0.001790 |
| `hashlib.md5` | 0.007207 |
| `hashlib.sha1` | 0.006742 |
| `hashlib.sha256` | 0.007733 |
| `hashlib.sha512` | 0.009293 |
| `binascii.crc32` | 0.002447 |

<details>
<summary>timeit benchmarking</summary>

```python
>>> import timeit
>>> import hashlib
>>> study_name = "tmp"
>>> trial_number = 1
>>> timeit.timeit(lambda: hash(f"{study_name}_{trial_number}"), number=10000)
>>> def _hash(s, f):
...     return int.from_bytes(f(s.encode()).digest(), byteorder="big")
>>> timeit.timeit(lambda: _hash(f"{study_name}_{trial_number}", hashlib.md5), number=10000)
>>> timeit.timeit(lambda: _hash(f"{study_name}_{trial_number}", hashlib.sha1), number=10000)
>>> timeit.timeit(lambda: _hash(f"{study_name}_{trial_number}", hashlib.sha256), number=10000)
>>> timeit.timeit(lambda: _hash(f"{study_name}_{trial_number}", hashlib.sha512), number=10000)
>>> timeit.timeit(lambda: binascii.crc32(f"{study_name}_{trial_number}".encode()), number=10000)
```
</details>

**Comment by hvy** [comment]
You're right, and a simple mod might also suffice. The logical change is probably a matter of a few lines of code (if not one), but we probably want to take a benchmark during the course of this PR, whichever approach we opt for. If a contributor is willing to work on this issue, a core maintainer could perhaps step in and help with setting up the benchmark, or alternatively, run the benchmark for him/her.

**Comment by avats-dev** [comment]
As we don't need a cryptographic hash and `crc32` results in an unsigned 32-bit int, it is a viable option here instead of `hash`. Also, I want to work on this but I don't have much idea about benchmarking. Also, do I need to raise multiple PRs with varying changes like one with just mod, one with `zlib.crc32` and then some with `hashlib.sha*` to test which one is best here? I am guessing there is some way by which I can test out which one is better for most scenarios locally and raise PR with that one specifically.

**Comment by HideakiImamura** [comment]
Thanks @avats-dev for your interest in contributing! For the benchmarking, I share some information.
- Optuna committers often benchmark with a benchmarking software called [kurobako](https://github.com/optuna/kurobako). It includes some hyperparameter optimization benchmark tasks.
- It takes a lot of time (20~30 hours) to perform dozens of runs on each of many datasets to perform a fair validation.

**Comment by hvy** [comment]
Right, so benchmarking with kurobako in a multi-process setting is not straight forward. One compromise could be to consider this a bug fix and relax the benchmarking requirements. Perhaps just comparing a single multi-process scenario between the current master and the fixed branch (to make sure that there are no obvious regressions)?

**Comment by sile** [comment]
> benchmarking with kurobako in a multi-process setting is not straight forward

Yes, I think so.

**Comment by avats-dev** [comment]
So, should I just move forward with `crc32` and implement it without benchmarking? Also, `crc32` is faster for short-length strings, so assuming it to be fastest might be misleading. [Ref](https://stackoverflow.com/questions/3665247/fastest-hash-for-non-cryptographic-uses)
Also, we can use [`xxhash`](https://github.com/Cyan4973/xxHash) or `murmer`  which are built specifically for these purposes and are quite fast. These could be implemented using [this](https://github.com/flier/pyfasthash) python library.

**Comment by hvy** [comment]
Thanks for the reference and the careful consideration. As it's also stated there, benchmarking is important, and in my understanding @toshihikoyanase addressed that in https://github.com/optuna/optuna/issues/3083#issuecomment-963833619 so certain degree. I did some quick comparisons in addition to that with `hash`, `crc32` and `sha512` using some examples in `optuna-examples`, which seemed consistent with those results. 

I'm a bit skeptical about introducing additional dependencies for this sole purpose. We may stay open for additional speed improvements on requests but I'm a bit worried that it's too much over-engineering.

**Comment by not522** [comment]
Resolved by #4131.
